File: Problems/373. Find K Pairs with Smallest Sums.py
"""
Problem Link: https://leetcode.com/problems/find-k-pairs-with-smallest-sums/
"""

import heapq
from typing import List


class Solution:
    def kSmallestPairs(self, nums1: List[int], nums2: List[int], k: int) -> List[List[int]]:
        if not nums1 or not nums2:
            return []

        heap = []
        result = []

        for i in range(min(k, len(nums1))):
            heapq.heappush(heap, (nums1[i] + nums2[0], nums1[i], nums2[0], 0))

        while heap and k > 0:
            _, num1, num2, idx = heapq.heappop(heap)
            result.append([num1, num2])
            if idx + 1 < len(nums2):
                heapq.heappush(
                    heap, (num1 + nums2[idx + 1], num1, nums2[idx + 1], idx + 1))
            k -= 1

        return result

File: Problems/373. Find K Pairs with Smallest Sums.java
/*
Problem Link: https://leetcode.com/problems/find-k-pairs-with-smallest-sums/
*/

import java.util.*;

class Solution {
    public List<List<Integer>> kSmallestPairs(int[] nums1, int[] nums2, int k) {

        List<List<Integer>> res = new ArrayList<>();
        PriorityQueue<int[]> pq = new PriorityQueue<>((a, b) -> (a[0] + a[1]) - (b[0] + b[1]));

        if (nums1.length == 0 || nums2.length == 0 || k == 0)
            return res;

        for (int i = 0; i < nums1.length && i < k; i++)
            pq.offer(new int[] { nums1[i], nums2[0], 0 });

        while (k-- > 0 && !pq.isEmpty()) {
            int[] curr = pq.poll();
            res.add(Arrays.asList(curr[0], curr[1]));
            if (curr[2] == nums2.length - 1)
                continue;
            pq.offer(new int[] { curr[0], nums2[curr[2] + 1], curr[2] + 1 });
        }
        return res;
    }
}